js/collection.js

```diff
@@ -0,0 +1,158 @@
+import { renderCards } from "./cards.js";
+
+const RARITIES = [
+  "Element",
+  "Common",
+  "Uncommon",
+  "Rare",
+  "HoloRare",
+  "UltraRare",
+  "SecretRare",
+];
+const PAGINATION_BTNS = {
+  container: document.getElementById("collection-pagination"),
+  previous: document.getElementById("pagination-previous"),
+  current: document.getElementById("pagination-current"),
+  next: document.getElementById("pagination-next"),
+};
+const COLLECTED_CARDS_NUMBER = document.getElementById(
+  "collected-cards-number"
+);
+
+//Object to store all supported sorting functions.
+//These are stored as closures to support reverse order without declaring new functions.
+//Can be used by calling:
+//array.sort(sort_functions[key](1, -1))
+//If you want to reverse order, swap the 1 and -1 when calling the function.
+const sort_functions = {
+  "Collector Number":
+    (before = 1, after = -1) =>
+    (a, b) => {
+      return parseInt(a["Collector Number"]) > parseInt(b["Collector Number"])
+        ? before
+        : after;
+    },
+  Rarity:
+    (before = 1, after = -1) =>
+    (a, b) => {
+      if (a["Rarity Folder"] === b["Rarity Folder"]) {
+        return sort_functions["Collector Number"]()(a, b);
+      } else {
+        return RARITIES.indexOf(a["Rarity Folder"]) <
+          RARITIES.indexOf(b["Rarity Folder"])
+          ? before
+          : after;
+      }
+    },
+};
+
+function getOwnedCards(cards_data) {
+  let ownedCards = [];
+  for (let item in localStorage) {
+    if (item.slice(0, 4) === "card") {
+      let card_id = item.split("-")[1];
+      ownedCards.push(
+        cards_data.find((card) => card["Collector Number"] === card_id)
+      );
+    }
+  }
+  return ownedCards;
+}
+
+function sortCards(cards, sortType, reverse = false) {
+  sortType = sortType ?? "Collector Number";
+  let before = reverse ? -1 : 1;
+  let after = reverse ? 1 : -1;
+  cards.sort(sort_functions[sortType](before, after));
+  return cards;
+}
+
+function sortAndFilterCards(cards_data, sort, reverse, fullCollection) {
+  if (fullCollection) {
+    return sortCards(cards_data, sort, reverse);
+  } else {
+    let cards = getOwnedCards(cards_data);
+    if (cards.length == 0) {
+      return null;
+    } else {
+      return sortCards(cards, sort, reverse);
+    }
+  }
+}
+
+//The hide parameter is boolean, so you can call it with an expression to use the function
+//in the same way you would use an if statement.
+function showOrHidePaginationControl(control, onclick, hide = false) {
+  if (hide) {
+    control.classList.add("hidden");
+  } else {
+    control.classList.remove("hidden");
+    control.onclick = onclick;
+  }
+}
+
+//If we will show more cards than the amount specified in cards_per_page,
+//then we split them into parts and show the pagination controls.
+//Note: If cards_per_page is set to "all", the comparison casts it to NaN, and as such it will always return false.
+function paginateCards(cards, cards_per_page, page, navigationFunction) {
```

Comment:
could you comment the shape of what this function returns and a summary of what the function does, ie
"Returns cards_per_page number of card data objects to render on the specified page. Additionally, wires and shows/ hides the pagination controls accordingly"
Took me a while to figure this out :')

Reply:
Done o7. The function looks more complex than it really is, but hopefully the comments help in clarifying how it works.